Side navigation (#1349)
Orange-OpenSource/Orange-Boosted-Bootstrap: site/content/docs/5.3/components/side-navigation.md

---
layout: docs
title: Side Navigation
description: Documentation and examples for Boosted's exclusive Brand responsive side navigation.
group: components
aliases:
  - "/docs/components/side-navigation/"
toc: true
added: "5.3"
---

## How it works

> Our side navigation is basically a `<nav>` that provides a hierarchy between links.

Reviewer: There is always a hierarchy? No need to use it juste for one level of links?

> Side navigation should be used inside an offcanvas using `.offcanvas-*` for responsive purpose. By default, we use `.offcanvas-lg` in our docs. Please refer to [Offcanvas]({{< docsref "/components/offcanvas#responsive" >}}) to learn more about its behavior. Since it uses [responsive offcanvas]({{< docsref "/components/offcanvas#responsive" >}}), please resize your browser to see its responsive behavior in action.

Reviewer: 'should' or 'must'?

Author: Should since we don't have any spec for responsive right now. Furthermore, it strongly depends on the people needs, so I'd let it as-is

{{< callout warning >}}
In order to make side navigation work properly, please make sure that the Javascript files from [Collapse]({{< docsref "/components/collapse#usage" >}}) and [Tabs]({{< docsref "/components/navs-tabs#javascript-behavior" >}}) are imported in your project.
{{< /callout >}}

> The side navigation provides 3 ready-to-use fully responsive themes. For each theme, we provide:

Reviewer: Not sure about the word 'theme', it sounds like the future 'color themes' and it may be confusing 🤔

- `.side-nav` to set up at the root element.
- `.side-nav-dark` to darken our theme variants.
- `.side-nav-scrollable` to ensure correct behavior of scroll boxes.
- `.side-nav-item` to ensure correct size and behavior of displayed elements.
- optional `.side-nav-static` to make collapsible side navigation static.
- optional `.side-nav-toggle` to embed a version of the collapse toggler.

> - optional `.side-nav-content` to make sure the content is well sized even with a toggler button.

Author: Nope but it depends on the variant. For example the accordion variant doesn't have the toggler button. I should group it with the line above maybe.

With the following structure:

```html
<nav class="side-nav side-nav-dark">
  <div class="side-nav-theme side-nav-static">
    <!-- If collapsible -->
    <button class="side-nav-toggle"></button>
    <div class="side-nav-content">
    <!-- End collapsible -->
      <a href="#" class="side-nav-item">Links</a>
    </div>
  </div>
</nav>
```

## Examples

> Side navigation come with 3 ready-to-use fully responsive themes. If these available themes don’t meet your specific needs, feel free to develop your own custom theme accordingly to the Orange Design System.

Reviewer: `don’t` or `don't`?

> Side navigation come with 3 ready-to-use fully responsive themes. If these available themes don’t meet your specific needs, feel free to develop your own custom theme accordingly to the Orange Design System.

Reviewer: 'Side navigation comes'?

Since most of side navigation we display use the responsive version of this component, please resize your browser to see the behavior.

{{< callout info >}}

> You can also notice that every example is resizable in order to show the behavior it has in different configurations.

Reviewer: Not sure if it need to be in a callout? Or if the callout should also include line 50?

          <div class="accordion-body">
            <ul class="list-unstyled m-0">
              <li><a href="#" class="side-nav-item">Label</a></li>
              <li><a href="#" class="side-nav-item">Label</a></li>
              <li><a href="#" class="side-nav-item">Label</a></li>
            </ul>
          </div>
        </div>
      </div>
      <div class="accordion-item">
        <a href="#" class="side-nav-item fw-bold">Financial</a>
      </div>
    </div>
  </div>
</nav>

<div class="d-flex flex-grow-1 p-2">Content that should fill the remaining space.</div>
{{< /example >}}

### With accordions

Here is the `.side-nav-accordion` theme. It should be used with [`.side-nav-static`](#static) since it doesn't have any collapse behavior. The whole accordion should be scrollable.

{{< example class="p-lg-0 side-nav-example d-block d-lg-flex" >}}
<button class="btn btn-primary d-lg-none" data-bs-target="#sideNavOffcanvas" data-bs-toggle="offcanvas" aria-haspopup="true" aria-controls="sideNavOffcanvas" id="sideNavToggle">Show basic side navigation</button>

> <div class="offcanvas-lg offcanvas-start h-100" role="region" aria-labelledby="sideNavToggle" id="sideNavOffcanvas">

Reviewer: Could be it confusing to use the `.h-100` utility class in some examples and to use a CSS "trick to remove `.h-100` inside the markup" in the others?

Author: This `.h-100` is outside the component so I couldn't replace it easily by CSS trick, I could study the impact inside offcanvas if needed.

                </svg>
                Charts
              </button>
            </h3>
            <div id="chartsCollapseIconsOc" class="accordion-collapse collapse" aria-labelledby="chartsHeaderIconsOc" data-bs-parent="#sideNavAccordionIconsOc">
              <div class="accordion-body">
                <ul class="list-unstyled m-0">
                  <li><a href="#" class="side-nav-item">Label</a></li>
                  <li><a href="#" class="side-nav-item">Label</a></li>
                  <li><a href="#" class="side-nav-item">Label</a></li>
                </ul>
              </div>
            </div>
          </div>
          <div class="accordion-item">
            <a href="#" class="side-nav-item fw-bold">
              <svg width="1.5rem" height="1.5rem" aria-hidden="true" focusable="false">
                <use xlink:href="/docs/{{< param docs_version >}}/assets/img/boosted-sprite.svg#financial"/>
              </svg>
              Financial
            </a>
          </div>
        </div>
      </div>
    </nav>

  </div>
</div>

<div class="d-none d-lg-flex flex-grow-1 p-2">Content that should fill the remaining space.</div>
{{< /example >}}

### Collapsible

> Here is the `.side-nav-collapsible` theme. The collapsible side navigation we provide uses [Tooltips]({{< docsref "/components/tooltips#usage" >}}) internally. It also uses a small trick to nicely collapse the side navigation with an empty `div`. (might change if we try to change the collapse thing on Bootstrap)

Reviewer: I would not say that it may change. Or I would at least rephrase it.

Author: Ah yeah it was a note for myself, sorry for that 🤣 Please let this one open to remember about it !

> Here is the `.side-nav-collapsible` theme. The collapsible side navigation we provide uses [Tooltips]({{< docsref "/components/tooltips#usage" >}}) internally. It also uses a small trick to nicely collapse the side navigation with an empty `div`. (might change if we try to change the collapse thing on Bootstrap)

Reviewer: How does it looks like with the new version of the tooltips?

Author: It should be even closer to the design! But I admit I haven't tried

                <use xlink:href="/docs/{{< param docs_version >}}/assets/img/boosted-sprite.svg#dashboard"/>
              </svg>
              Dashboard
            </a>
          </li>
          <li data-bs-toggle="tooltip" data-bs-title="Charts" data-bs-placement="right">
            <a href="#" class="side-nav-item fw-bold">
              <svg width="1.5rem" height="1.5rem" aria-hidden="true" focusable="false">
                <use xlink:href="/docs/{{< param docs_version >}}/assets/img/boosted-sprite.svg#chart"/>
              </svg>
              Charts
            </a>
          </li>
          <li data-bs-toggle="tooltip" data-bs-title="Financial" data-bs-placement="right">
            <a href="#" class="side-nav-item fw-bold">
              <svg width="1.5rem" height="1.5rem" aria-hidden="true" focusable="false">
                <use xlink:href="/docs/{{< param docs_version >}}/assets/img/boosted-sprite.svg#financial"/>
              </svg>
              Financial
            </a>
          </li>
        </ul>
      </div>
    </nav>

  </div>
</div>

<div class="d-none d-lg-flex flex-grow-1 p-2">Content that should fill the remaining space.</div>
{{< /example >}}

> ### Collapsible with drawer content

Reviewer: Should it be possible to close the drawer using `Esc` key? Maybe as an option?

Author: It'll come with the Js and the returns of a11y experts !

Here is the `.side-nav-content-with-drawer` theme. The collapsible side navigation we provide uses [Tooltips]({{< docsref "/components/tooltips#usage" >}}) internally.

The collapsible side navigation with drawer content comes with ready to use `onclick` calls:
- On the Offcanvas toggler button to `show` its content.
- On the Offcanvas close button to `hide` its content.

> - On each menu item to `show` its content. (nav-item)

Reviewer: `its content (nav-item).` maybe?

                  <use xlink:href="/docs/{{< param docs_version >}}/assets/img/boosted-sprite.svg#dashboard"/>
                </svg>
                <span class="visually-hidden">Open dashboard tab</span>
              </button>
            </li>
            <li class="nav-item" role="none" data-bs-toggle="tooltip" data-bs-title="Charts" data-bs-placement="right">
              <button class="side-nav-item" data-bs-toggle="tab" data-bs-target="#chartsTab1" type="button" role="menuitem" aria-controls="chartsTab1" aria-expanded="false" aria-haspopup="true" onclick="boosted.Collapse.getOrCreateInstance('#sideNavCollapseCwdc1').show()">
                <svg class="align-self-center" width="1.5rem" height="1.5rem" aria-hidden="true" focusable="false">
                  <use xlink:href="/docs/{{< param docs_version >}}/assets/img/boosted-sprite.svg#chart"/>
                </svg>
                <span class="visually-hidden">Open charts tab</span>
              </button>
            </li>
            <li class="nav-item" role="none" data-bs-toggle="tooltip" data-bs-title="Financial" data-bs-placement="right">

>               <button class="side-nav-item" data-bs-toggle="tab" data-bs-target="#financialTab1" type="button" role="menuitem" aria-controls="financialTab1" aria-expanded="false" aria-haspopup="true" onclick="boosted.Collapse.getOrCreateInstance('#sideNavCollapseCwdc1').show()">

Reviewer: Remove inline JS.

Reviewer: Not sure about my comment, it may be the easiest way to explain the code example? I need to think about it... 🤔

Author: I need to add those until we have some proper Js otherwise, we wouldn't see the expected behavior of this component. As said in design review, depending on the a11y returns, the Js might be done in a standalone PR after to ease the review process.

<div class="offcanvas-lg offcanvas-start" role="region" aria-labelledby="sideNavCollapsibleStaticToggle" id="sideNavCollapsibleStatic">
  <div class="offcanvas-header p-0">
    <button type="button" class="btn-close ms-auto" data-bs-target="#sideNavCollapsibleStatic" data-bs-dismiss="offcanvas" aria-expanded="true" aria-controls="sideNavCollapsibleStatic"><span class="visually-hidden">Close</span></button>
  </div>
  <div class="offcanvas-body h-100 p-0">
    <nav class="side-nav side-nav-static" aria-label="Static collapsible side navigation inside an offcanvas">
      <div class="side-nav-collapsible">
        ...
      </div>
    </nav>
  </div>
</div>

<div class="d-none d-lg-flex flex-grow-1 p-2">Content that should fill the remaining space.</div>
```

## Scrollable

Even if we handle the scroll in a particular way inside our predefined themes, you may want to change this behavior with the `.side-nav-scrollable`. If you do so, please make sure that every part is accessible through scroll (the icon part and the content one).

<!-- Following the spec review

> Here are some examples with the [collapsible with drawer content](#collapsible-with-drawer-content), that default to a scroll on both parts (the icon part and the content part) excluding the title.

Reviewer: Note sure about why this part is inside a big comment...

Author: It's due to a spec review we had, removing it for sure

  <div class="offcanvas-start offcanvas-lg d-flex" role="region" aria-labelledby="sideNavCwdc2Toggle" id="sideNavCwdc2">
    <div class="offcanvas-header p-0 side-nav-content-with-drawer">
      <button type="button" class="btn-close ms-auto" data-bs-dismiss="offcanvas" data-bs-target="#sideNavCwdc2" aria-expanded="true" aria-controls="sideNavCwdc2"><span class="visually-hidden">Close</span></button>
    </div>
    <div class="offcanvas-body p-0">
      <nav class="side-nav" aria-label="Collapsible side navigation with drawer content with scroll boxes including the title">
        <div class="side-nav-content-with-drawer">
          <button class="side-nav-toggle collapsed" data-bs-toggle="collapse" data-bs-target="#sideNavCollapseCwdc2"><span class="visually-hidden">Open/Close side navigation</span></button>
          <div class="side-nav-content">
            <ul class="h-100 list-unstyled m-0 side-nav-dark side-nav-scrollable" role="tablist">
              <li role="presentation" data-bs-toggle="tooltip" data-bs-title="Dashboard" data-bs-placement="right">

>                 <button class="side-nav-item active" data-bs-toggle="tab" data-bs-target="#dashboardTab2" type="button" role="tab" aria-controls="dashboardTab2" aria-selected="true" onclick="boosted.Collapse.getOrCreateInstance('#sideNavCollapseCwdc2').show()">

Reviewer: It seems that `aria-selected="<true | false >"` attribute is not allowed on a `button` element (https://www.w3.org/TR/html-aria/#docconformance).
You may need to double check with our expert and maybe make some changes.

Author: It's done automatically by Bootstrap Js, it might change if we have specific Js for this one.
Meanwhile, do you think it would be better to use `<a>`? I don't know what would be the great HTML here. I feel like the links are overkill since we don't change the page but I might be wrong.

                </svg>
                Financial
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div class="d-none d-lg-flex flex-grow-1 p-2">Content that should fill the remaining space.</div>
</div>

```html
<button class="btn btn-primary d-lg-none" data-bs-target="#sideNavCollapsibleRight" data-bs-toggle="offcanvas" aria-haspopup="true" aria-controls="sideNavCollapsibleRight" id="sideNavCollapsibleRightToggle">Collapsing side navigation</button>

<div class="offcanvas-lg offcanvas-end" role="region" aria-labelledby="sideNavCollapsibleRightToggle" id="sideNavCollapsibleRight">
  ...
</div>

<div class="d-none d-lg-flex flex-grow-1 p-2">Content that should fill the remaining space.</div>
```
-->

## Dark variant

Add `.side-nav-dark` to the `.side-nav` or its children for a dark variant. For example the [collapsible side navigation with drawer content](#collapsible-with-drawer-content) is made with a part of dark variant.

We strongly recommend to use `.offcanvas-dark` (resp. `.btn-close-white`) on top of `.offcanvas-*` (resp. `.btn-close`) in order to have the correct design.

{{< callout info >}}

> Dark variant can be applied on root of any example except the [collapsible side navigation with drawer content](#collapsible-with-drawer-content) one.

Reviewer: It may be a bit weird to give an example with a link to collapsible side navigation with drawer content and add a callout 5 lines after saying that it's not compatible with collapsible side navigation with drawer content.

Author: ```suggestion
Dark variant shouldn't be applied on the root element of the [collapsible side navigation with drawer content](#collapsible-with-drawer-content) one.
```
Would it be better here ?

Reviewer: Yes, I prefer this version 👍
